### PR TITLE
[B+C] Implement Dropper interface. Adds BUKKIT-3750

### PR DESCRIPTION
**Pull request by Xephi**
The Issue:

Actually , there is no way to get the Dropper inventory with the simple Bukkit API , and you can't cast the BlockDispenseEvent block.getState() to a Dispenser cause that return a CraftBlockState

Justification for this PR:
This PR add an Interface for Dropper that create a correct CraftDropper extends CraftBlockState

PR Breakdown:

This PR implement the Dropper interface with create is own CraftBlockState like with a Dispenser

Relevant PR:

CraftBukkit: https://github.com/Bukkit/CraftBukkit/pull/1071

Jira Ticket :
Leaky: https://bukkit.atlassian.net/browse/BUKKIT-3750

### CODE REVIEW

File: src/main/java/org/bukkit/block/Dropper.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ */
+public interface Dropper extends Dispenser {
+
+    /**
+     * Attempts to dispense the contents of this block
+     *
+     * If the block is no longer a dropper, this will return false
+     *
+     * @return true if successful, otherwise false
+     */
```
Comment (natemort) [inline src/main/java/org/bukkit/block/Dropper.java:23]: I find this javadoc to be pretty non-descriptive, and looks like almost an exact copy of Dispenser#dispense()'s javadocs.  Perhaps more information can be provided regarding what this actually does.

**Comment by natemort** [comment]
@Xephi See the second half of https://github.com/Bukkit/CraftBukkit/pull/1071#issuecomment-15153264 and please apply it to this PR as well.

File: src/main/java/org/bukkit/block/Dropper.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package org.bukkit.block;
+/**
+ * Represents a dropper.
+ */
+public interface Dropper extends Dispenser {
+
+    /**
+     * Try to drop one of item from one itemstack the dropper contain
+     *
+     * This dispense method return false if the block is no longer a dropper
```
Comment (natemort) [inline src/main/java/org/bukkit/block/Dropper.java]: This sounds like it should be moved down to @return  .

File: src/main/java/org/bukkit/block/Dropper.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.bukkit.block;
+/**
+ * Represents a dropper.
+ */
+public interface Dropper extends Dispenser {
+
+    /**
+     * Try to drop one of item from one itemstack the dropper contain
```
Comment (natemort) [inline src/main/java/org/bukkit/block/Dropper.java:8]: How is that item determined?

Comment (Xephi) [inline src/main/java/org/bukkit/block/Dropper.java:8]: I've added this to the javadoc

**Comment by natemort** [comment]
@Xephi I understand you've spent a lot of time working on this PR, but I've talked with the Core Team and there's some changes we'd like to see take place:
1.  We feel that Dropper should not extend Dispenser.  I understand we made our previous comments regarding simply using override, but we feel it would be best for this API.  We also understand that in nms(net.minecraft.server), that it extends Dispenser, but the bukkit API should not necessarily be a direct reflection.  We apologize for any time spent working on aspects that will not be pulled and hope that you'll continue working on this
2.  Change Dropper#dispense() to Dropper#drop() and document in great detail the behavior of a Dropper.  Explain in detail how it sometimes acts like a hopper, and sometimes acts like a Dispenser.

Once again, we apologize for time you've spent on aspects of this that will not be pulled, but we appreciate the great amount of work you're going through to try and improve the Bukkit project.

**Comment by Xephi** [comment]
That's what i've done on the first time , no problem , i just need to revert my changes on Eclipse and it's ok :') i pull all my change when i can , sorry for probably delay

Xephi

**Comment by Xephi** [comment]
@evilmidget38 Huuum do i have to close this maybe and PR again with change ? i think that could be better ?

File: src/main/java/org/bukkit/block/Dropper.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+     * 
+     * Example : 
+     * 
+     * 1 2 3
+     * 4   6
+     * 7
+     * 
+     * Dropper can drop one by one these items : 1 3 2 7 1 2 2 3 4 6 6 ...
+     * 
+     * If a container is around the dropper, this item go inside (like with a Hooper)
```
Comment (natemort) [inline src/main/java/org/bukkit/block/Dropper.java]: Should be "Hopper"

**Comment by Xephi** [comment]
CraftBukkit PR updated to 1.5.1 BlockDropper version

Xephi

**Comment by Xephi** [comment]
I'll close this and PR again

Xephi

**Pull request closed without merging**